diff --git a/config.h b/config.h
index ce0d1d4..c2346c1 100644
--- a/config.h
+++ b/config.h
@@ -37,17 +37,11 @@ static const uint8_t borders[] = {3,5,5,4};
 #define LOOK_INTO "WM_NAME"
- static const char *ignore_names[] = {"bar", "xclock"};
+ static const char *ignore_names[] = {"lemonbar", "bar", "xclock"};
 ///--Menus and Programs---///
-static const char *menucmd[]   = { "my_menu.sh", NULL };
-static const char *gmrun[]     = { "my_menu2.sh",NULL};
-static const char *terminal[]  = { "urxvtc", NULL };
+static const char *menucmd[]   = { "dmenu_run", NULL };
+static const char *terminal[]  = { "termite", NULL };
 static const char *click1[]    = { "xdotool","click", "1", NULL };
 static const char *click2[]    = { "xdotool","click", "2", NULL };
 static const char *click3[]    = { "xdotool","click", "3", NULL };
-static const char *vol_up[]    = { "pamixer", "-u", "-i", "3", "--allow-boost", NULL };
-static const char *vol_down[]  = { "pamixer", "-u", "-d", "3", "--allow-boost", NULL };
-static const char *vol_mute[]  = { "amixer", "set", "Master", "mute", "-q", NULL };
-static const char *bright_up[]  = { "light", "-A", "5", NULL };
-static const char *bright_down[]  = { "light", "-U", "5", NULL };
 ///--Custom foo---///
 static void halfandcentered(const Arg *arg)
 {
@@ -186,8 +180,7 @@ static key keys[] = {
     {  MOD |SHIFT,        XK_Left,       cursor_move,       {.i=TWOBWM_CURSOR_LEFT}},
     // Start programs
     {  MOD ,              XK_Return,     start,             {.com = terminal}},
-    {  MOD ,              XK_w,          start,             {.com = menucmd}},
-    {  MOD |SHIFT,        XK_w,          start,             {.com = gmrun}},
+    {  MOD ,              XK_d,          start,             {.com = menucmd}},
     // Exit or restart 2bwm
     {  MOD |CONTROL,      XK_q,          twobwm_exit,       {.i=0}},
     {  MOD |CONTROL,      XK_r,          twobwm_restart,    {.i=0}},
@@ -195,12 +188,7 @@ static key keys[] = {
     // Fake clicks using xdotool
     {  MOD |CONTROL,      XK_Up,         start,             {.com = click1}},
     {  MOD |CONTROL,      XK_Down,       start,             {.com = click2}},
-	{  MOD |CONTROL,      XK_Right,      start,             {.com = click3}},
-    {  0x000000,          0x1008ff13, start,             {.com = vol_up}},
-    {  0x000000,          0x1008ff11,  start,             {.com = vol_down}},
-    {  0x000000,          0x1008ff15, start,             {.com = vol_mute}},
-    {  0x000000,          0x1008ff02, start,             {.com = bright_up}},
-    {  0x000000,          0x1008ff03,  start,             {.com = bright_down}},
+    {  MOD |CONTROL,      XK_Right,      start,             {.com = click3}},
     // Change current workspace
        DESKTOPCHANGE(     XK_1,                             0)
        DESKTOPCHANGE(     XK_2,                             1)
