diff --git a/block/Kconfig.iosched b/block/Kconfig.iosched
index fd2cefa47d35..4dbabfb5265a 100644
--- a/block/Kconfig.iosched
+++ b/block/Kconfig.iosched
@@ -56,6 +56,9 @@ choice
 	config DEFAULT_NOOP
 		bool "No-op"
 
+	config DEFAULT_BFQ
+		bool "BFQ" if IOSCHED_BFQ=y
+
 endchoice
 
 config DEFAULT_IOSCHED
@@ -63,6 +66,7 @@ config DEFAULT_IOSCHED
 	default "deadline" if DEFAULT_DEADLINE
 	default "cfq" if DEFAULT_CFQ
 	default "noop" if DEFAULT_NOOP
+    default "bfq" if DEFAULT_BFQ
 
 config MQ_IOSCHED_DEADLINE
 	tristate "MQ deadline I/O scheduler"
