diff --git a/src/tools/toolfactory.cpp b/src/tools/toolfactory.cpp
index 35212403..8fd8be96 100644
--- a/src/tools/toolfactory.cpp
+++ b/src/tools/toolfactory.cpp
@@ -53,13 +53,9 @@ CaptureTool* ToolFactory::CreateTool(CaptureTool::Type t, QObject* parent)
         if_TYPE_return_TOOL(TYPE_EXIT, ExitTool);
 #ifdef ENABLE_IMGUR
         if_TYPE_return_TOOL(TYPE_IMAGEUPLOADER, ImgUploaderTool);
-#endif
-#if !defined(Q_OS_MACOS)
-        if_TYPE_return_TOOL(TYPE_OPEN_APP, AppLauncher);
 #endif
         if_TYPE_return_TOOL(TYPE_PIXELATE, PixelateTool);
         if_TYPE_return_TOOL(TYPE_REDO, RedoTool);
-        if_TYPE_return_TOOL(TYPE_PIN, PinTool);
         if_TYPE_return_TOOL(TYPE_TEXT, TextTool);
         if_TYPE_return_TOOL(TYPE_CIRCLECOUNT, CircleCountTool);
         if_TYPE_return_TOOL(TYPE_SIZEINCREASE, SizeIncreaseTool);
diff --git a/src/utils/confighandler.cpp b/src/utils/confighandler.cpp
index e202e5dc..7aa380d3 100644
--- a/src/utils/confighandler.cpp
+++ b/src/utils/confighandler.cpp
@@ -100,7 +100,7 @@ static QMap<class QString, QSharedPointer<ValueHandler>>
     OPTION("useJpgForClipboard"          ,Bool               ( false         )),
     OPTION("uploadWithoutConfirmation"   ,Bool               ( false         )),
     OPTION("saveAfterCopy"               ,Bool               ( false         )),
-    OPTION("savePath"                    ,ExistingDir        (               )),
+    OPTION("savePath"                    ,String             ( ""            )),
     OPTION("savePathFixed"               ,Bool               ( false         )),
     OPTION("saveAsFileExtension"         ,SaveFileExtension  (               )),
     OPTION("saveLastRegion"              ,Bool               ( false         )),
@@ -158,9 +158,6 @@ static QMap<QString, QSharedPointer<KeySequence>> recognizedShortcuts = {
     SHORTCUT("TYPE_CANCEL"              ,   "Ctrl+Backspace"        ),
 #ifdef ENABLE_IMGUR
     SHORTCUT("TYPE_IMAGEUPLOADER"       ,                           ),
-#endif
-#if !defined(Q_OS_MACOS)
-    SHORTCUT("TYPE_OPEN_APP"            ,   "Ctrl+O"                ),
 #endif
     SHORTCUT("TYPE_PIXELATE"            ,   "B"                     ),
     SHORTCUT("TYPE_INVERT"              ,   "I"                     ),
@@ -189,7 +186,6 @@ static QMap<QString, QSharedPointer<KeySequence>> recognizedShortcuts = {
 #else
     SHORTCUT("TYPE_DELETE_CURRENT_TOOL" ,   "Delete"                ),
 #endif
-    SHORTCUT("TYPE_PIN"                 ,                           ),
     SHORTCUT("TYPE_SIZEINCREASE"        ,                           ),
     SHORTCUT("TYPE_SIZEDECREASE"        ,                           ),
     SHORTCUT("TYPE_CIRCLECOUNT"         ,                           ),
@@ -199,10 +195,8 @@ static QMap<QString, QSharedPointer<KeySequence>> recognizedShortcuts = {
 // CLASS CONFIGHANDLER
 
 ConfigHandler::ConfigHandler()
-  : m_settings(QSettings::IniFormat,
-               QSettings::UserScope,
-               qApp->organizationName(),
-               qApp->applicationName())
+  : m_settings("@configFile@",
+               QSettings::IniFormat)
 {
     static bool firstInitialization = true;
     if (firstInitialization) {
diff --git a/src/utils/confighandler.h b/src/utils/confighandler.h
index ecc95665..7269f63d 100644
--- a/src/utils/confighandler.h
+++ b/src/utils/confighandler.h
@@ -8,6 +8,7 @@
 #include <QStringList>
 #include <QVariant>
 #include <QVector>
+#include <QDir>
 
 #define CONFIG_GROUP_GENERAL "General"
 #define CONFIG_GROUP_SHORTCUTS "Shortcuts"
@@ -67,12 +68,20 @@ public:
 
     static ConfigHandler* getInstance();
 
+    QString savePath() {
+        QString val = value(QStringLiteral("savePath")).toString()
+            .replace("$HOME", QDir::homePath());
+        QDir().mkpath(val);
+        return val;
+    }
+
+    CONFIG_SETTER(setSavePath, savePath, QString);
+
     // Definitions of getters and setters for config options
     // Some special cases are implemented regularly, without the macro
     // NOTE: When adding new options, make sure to add an entry in
     // recognizedGeneralOptions in the cpp file.
     CONFIG_GETTER_SETTER(userColors, setUserColors, QVector<QColor>);
-    CONFIG_GETTER_SETTER(savePath, setSavePath, QString)
     CONFIG_GETTER_SETTER(savePathFixed, setSavePathFixed, bool)
     CONFIG_GETTER_SETTER(uiLanguage, setUiLanguage, QString)
     CONFIG_GETTER_SETTER(uiColor, setUiColor, QColor)
diff --git a/src/widgets/capture/capturetoolbutton.cpp b/src/widgets/capture/capturetoolbutton.cpp
index f683f46b..658b0b39 100644
--- a/src/widgets/capture/capturetoolbutton.cpp
+++ b/src/widgets/capture/capturetoolbutton.cpp
@@ -145,13 +145,7 @@ static std::map<CaptureTool::Type, int> buttonTypeOrder
       { CaptureTool::TYPE_IMAGEUPLOADER, 17 },
 #endif
       { CaptureTool::TYPE_ACCEPT, 18 },
-#if !defined(Q_OS_MACOS)
-      { CaptureTool::TYPE_OPEN_APP, 19 }, { CaptureTool::TYPE_EXIT, 20 },
-      { CaptureTool::TYPE_PIN, 21 },
-#else
-      { CaptureTool::TYPE_EXIT, 19 }, { CaptureTool::TYPE_PIN, 20 },
-#endif
-
+      { CaptureTool::TYPE_EXIT, 20 },
       { CaptureTool::TYPE_SIZEINCREASE, 22 },
       { CaptureTool::TYPE_SIZEDECREASE, 23 },
 };
@@ -176,9 +170,6 @@ QList<CaptureTool::Type> CaptureToolButton::iterableButtonTypes = {
 #ifdef ENABLE_IMGUR
     CaptureTool::TYPE_IMAGEUPLOADER,
 #endif
-#if !defined(Q_OS_MACOS)
-    CaptureTool::TYPE_OPEN_APP,
-#endif
-    CaptureTool::TYPE_PIN,           CaptureTool::TYPE_SIZEINCREASE,
-    CaptureTool::TYPE_SIZEDECREASE,  CaptureTool::TYPE_ACCEPT,
+    CaptureTool::TYPE_SIZEINCREASE,  CaptureTool::TYPE_SIZEDECREASE,
+    CaptureTool::TYPE_ACCEPT,
 };
diff --git a/src/widgets/capture/capturewidget.cpp b/src/widgets/capture/capturewidget.cpp
index fd866786..489e1085 100644
--- a/src/widgets/capture/capturewidget.cpp
+++ b/src/widgets/capture/capturewidget.cpp
@@ -329,8 +329,6 @@ void CaptureWidget::initButtons()
 #ifdef ENABLE_IMGUR
             buttonList->removeOne(CaptureTool::TYPE_IMAGEUPLOADER);
 #endif
-            buttonList->removeOne(CaptureTool::TYPE_OPEN_APP);
-            buttonList->removeOne(CaptureTool::TYPE_PIN);
         }
     }
     QVector<CaptureToolButton*> vectorButtons;
