diff --git a/src/utils/confighandler.cpp b/src/utils/confighandler.cpp
index 6786225..3a55a73 100644
--- a/src/utils/confighandler.cpp
+++ b/src/utils/confighandler.cpp
@@ -16,9 +16,8 @@
 #endif
 
 ConfigHandler::ConfigHandler()
-{
-    m_settings.setDefaultFormat(QSettings::IniFormat);
-}
+  : m_settings("@configFile@", QSettings::IniFormat)
+{}
 
 QVector<CaptureToolButton::ButtonType> ConfigHandler::getButtons()
 {
@@ -49,11 +48,7 @@ QVector<CaptureToolButton::ButtonType> ConfigHandler::getButtons()
                 << CaptureToolButton::TYPE_UNDO << CaptureToolButton::TYPE_REDO
                 << CaptureToolButton::TYPE_COPY << CaptureToolButton::TYPE_SAVE
                 << CaptureToolButton::TYPE_EXIT
-                << CaptureToolButton::TYPE_IMAGEUPLOADER
-#if not defined(Q_OS_MACOS)
-                << CaptureToolButton::TYPE_OPEN_APP
-#endif
-                << CaptureToolButton::TYPE_PIN << CaptureToolButton::TYPE_TEXT
+                << CaptureToolButton::TYPE_TEXT
                 << CaptureToolButton::TYPE_CIRCLECOUNT;
     }
 
@@ -105,7 +100,8 @@ QVector<QColor> ConfigHandler::getUserColors()
 
 QString ConfigHandler::savePath()
 {
-    return m_settings.value(QStringLiteral("savePath")).toString();
+    return m_settings.value(QStringLiteral("savePath")).toString()
+        .replace("$HOME", QDir::homePath());
 }
 
 void ConfigHandler::setSavePath(const QString& savePath)
