diff --git a/src/music-players.h b/src/music-players.h
index a44460a..d05f52f 100644
--- a/src/music-players.h
+++ b/src/music-players.h
@@ -3102,6 +3102,16 @@
   { "Jolla", 0x2931, "Sailfish (ID1)", 0x0a01,
       DEVICE_FLAGS_ANDROID_BUGS },
 
+  /* In update 4 the order of devices was changed for
+     better OS X / Windows suport and another device-id
+     got assigned for the MTP */
+  { "Jolla", 0x2931, "Sailfish (ID2)", 0x0a05,
+      DEVICE_FLAGS_ANDROID_BUGS },
+
+  /* In a later version, the ID changed again. */
+  { "Jolla", 0x2931, "Sailfish (ID3)", 0x0a07,
+      DEVICE_FLAGS_ANDROID_BUGS },
+
   /*
    * TCL? Alcatel?
    */
@@ -3129,12 +3139,6 @@
   { "DigiLand", 0x1f3a, "DL701Q", 0x0c02,
       DEVICE_FLAGS_ANDROID_BUGS },
 
-  /* In update 4 the order of devices was changed for
-     better OS X / Windows suport and another device-id
-     got assigned for the MTP */
-  { "Jolla", 0x2931, "Sailfish (ID2)", 0x0a05,
-      DEVICE_FLAGS_ANDROID_BUGS },
-
   /*
    * bq
    * https://sourceforge.net/p/libmtp/feature-requests/128/
