--- a/drivers/usb/storage/uas.c	2016-04-03 21:45:04.326212268 +0200
+++ b/drivers/usb/storage/uas.c	2016-04-03 21:51:45.263114257 +0200
@@ -812,7 +812,7 @@
 	.slave_configure = uas_slave_configure,
 	.eh_abort_handler = uas_eh_abort_handler,
 	.eh_bus_reset_handler = uas_eh_bus_reset_handler,
-	.can_queue = 65536,	/* Is there a limit on the _host_ ? */
+	.can_queue = MAX_CMNDS,	/* Is there a limit on the _host_ ? */
 	.this_id = -1,
 	.sg_tablesize = SG_NONE,
 	.skip_settle_delay = 1,
