diff --git a/src/windows/pluginloader/pluginloader.c b/src/windows/pluginloader/pluginloader.c
index 9e8556f..c50be2a 100644
--- a/src/windows/pluginloader/pluginloader.c
+++ b/src/windows/pluginloader/pluginloader.c
@@ -1510,7 +1510,7 @@ void dispatcher(int functionid, Stack &stack){
 				NPObject *objectValue;
 				NPError result;
 
-				if (variable == NPPVpluginScriptableNPObject)
+				if (variable == NPPVpluginScriptableNPObject && pluginFuncs.getvalue)
 					result = pluginFuncs.getvalue(instance, variable, &objectValue);
 				else{
 					DBG_WARN("FUNCTION_NPP_GETVALUE_OBJECT - variable %d not allowed", variable);
