diff --git a/internal/action/list.go b/internal/action/list.go
index abfe0fda..e4269f67 100644
--- a/internal/action/list.go
+++ b/internal/action/list.go
@@ -9,7 +9,6 @@ import (
 	"os/exec"
 	"strings"
 
-	"github.com/fatih/color"
 	"github.com/gopasspw/gopass/internal/action/exit"
 	"github.com/gopasspw/gopass/internal/store/leaf"
 	"github.com/gopasspw/gopass/internal/tree"
@@ -117,7 +116,6 @@ func redirectPager(ctx context.Context, subtree *tree.Root) (io.Writer, *bytes.B
 	if pager := os.Getenv("PAGER"); pager == "" {
 		return stdout, nil
 	}
-	color.NoColor = true
 	buf := &bytes.Buffer{}
 	return buf, buf
 }
