diff --git a/pkg/action/list.go b/pkg/action/list.go
index 8a96407..45b416e 100644
--- a/pkg/action/list.go
+++ b/pkg/action/list.go
@@ -9,7 +9,6 @@ import (
 	"os/exec"
 	"strings"
 
-	"github.com/fatih/color"
 	"github.com/gopasspw/gopass/pkg/ctxutil"
 	"github.com/gopasspw/gopass/pkg/termutil"
 	"github.com/gopasspw/gopass/pkg/tree"
@@ -115,7 +114,6 @@ func redirectPager(ctx context.Context, subtree tree.Tree) (io.Writer, *bytes.Bu
 	if pager := os.Getenv("PAGER"); pager == "" {
 		return stdout, nil
 	}
-	color.NoColor = true
 	buf := &bytes.Buffer{}
 	return buf, buf
 }
