diff --git a/src/tools/rustfmt/src/config/mod.rs b/src/tools/rustfmt/src/config/mod.rs
index cd90e090..a6db6da1 100644
--- a/src/tools/rustfmt/src/config/mod.rs
+++ b/src/tools/rustfmt/src/config/mod.rs
@@ -282,7 +282,7 @@ impl Config {
                 }
             }
 
-            Ok(None)
+            Ok(Some(env!("DEFAULT_CONFIG_FILE").into()))
         }
 
         match resolve_project_file(dir)? {
