diff --git a/toolkit/components/enterprisepolicies/PoliciesHelpers.sys.mjs b/toolkit/components/enterprisepolicies/PoliciesHelpers.sys.mjs
index 22aeaf068590..19d59d3069d8 100644
--- a/toolkit/components/enterprisepolicies/PoliciesHelpers.sys.mjs
+++ b/toolkit/components/enterprisepolicies/PoliciesHelpers.sys.mjs
@@ -644,6 +644,7 @@ export function clearBlockedAboutPages() {
 }
 
 export function blockAboutPage(manager, feature) {
+  return;
   addChromeURLBlocker();
   gBlockedAboutPages.push(feature);
 
