diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 4579c6dd..d1d8e67 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -199,7 +199,7 @@ static inline void bfq_schedule_dispatch(struct bfq_data *bfqd)
 {
 	if (bfqd->queued != 0) {
 		bfq_log(bfqd, "schedule dispatch");
-		kblockd_schedule_work(bfqd->queue, &bfqd->unplug_work);
+		kblockd_schedule_work(&bfqd->unplug_work);
 	}
 }
 
