--- a/dlls/winscard/winscard.c	1970-01-01 01:00:01.000000000 +0100
+++ b/dlls/winscard/winscard.c	2016-06-06 01:52:53.631444433 +0200
@@ -1527,7 +1527,7 @@
 {
     LONG lRet;
     TRACE(" 0x%08X %p %p %p %p %p %p\n",(unsigned int) hCard,mszReaderNames,pcchReaderLen,pdwState,pdwProtocol,pbAtr,pcbAtrLen);
-    if(!pcchReaderLen || !pdwState || !pdwProtocol || !pcbAtrLen)
+    if(!pcchReaderLen || !pcbAtrLen)
         lRet = SCARD_E_INVALID_PARAMETER;
     else if(!liteSCardStatus)
         lRet = SCARD_F_INTERNAL_ERROR;
